### PR TITLE
Build names and links at runtime - no more sqlite

### LINKED ISSUE #17691: Creating fail with Could not find container for entity id <id> after upgrading to 1.9.0
```

Client:
 Version:      1.9.0
 API version:  1.21
 Go version:   go1.4.2
 Git commit:   76d6bc9
 Built:        Tue Nov  3 17:43:42 UTC 2015
 OS/Arch:      linux/amd64

Server:
 Version:      1.9.0
 API version:  1.21
 Go version:   go1.4.2
 Git commit:   76d6bc9
 Built:        Tue Nov  3 17:43:42 UTC 2015
 OS/Arch:      linux/amd64

Linux gt-xps 3.19.0-31-generic #36~14.04.1-Ubuntu SMP Thu Oct 8 10:21:08 UTC 2015 x86_64 x86_64 x86_64 GNU/Linux

Containers: 1
Images: 37
Server Version: 1.9.0
Storage Driver: overlay
 Backing Filesystem: extfs
Execution Driver: native-0.2
Logging Driver: json-file
Kernel Version: 3.19.0-31-generic
Operating System: Ubuntu 14.04.3 LTS
CPUs: 4
Total Memory: 7.372 GiB
Name: gt-xps
ID: 2562:2TQM:TR3I:CDFJ:T6JP:WXBN:AV4F:LZYT:HIAX:ZFYC:MMWD:Q7AN
Username: gionn
Registry: https://index.docker.io/v1/
WARNING: No swap limit support
```

Before I was using 1.8.3 with aufs storage driver.

```
$ docker run --rm -it --name cloudesire_broker_1 ubuntu
Error response from daemon: Could not find container for entity id 72845af4d9964d42eab55fd84c1a2cc25d7b32b8c587ed47b766d9a7a8fe5396

$ docker inspect cloudesire_broker_1
Error: No such image or container: cloudesire_broker_1
[]

$ docker inspect 72845af4d9964d42eab55fd84c1a2cc25d7b32b8c587ed47b766d9a7a8fe5396
Error: No such image or container: 72845af4d9964d42eab55fd84c1a2cc25d7b32b8c587ed47b766d9a7a8fe5396
[]

$ docker ps -a
CONTAINER ID        IMAGE               COMMAND             CREATED             STATUS              PORTS               NAMES
```

First debug happened in compose issue tracker since it appeared related to it: https://github.com/docker/compose/issues/2316

> I've a feeling that has something to do with containers that was existing before the upgrade to 1.9.0 from 1.8.3, with the contextual switch from storage driver aufs to overlay.

### PR DESCRIPTION
**Pull request by cpuguy83**
This stores all names and link info in memory at runtime rather than relying on sqlite.

Please test thoroughly.
I added some tests, but this is a big change!

Fixes #17691

### CODE REVIEW

**Comment by cpuguy83** [comment]
phew, these rebases are rough.

**Comment by cpuguy83** [comment]
And realizing that hostconfig.Links was being set to nil before, so this information is only in sqlite...

**Comment by LK4D4** [comment]
−158,972

File: hack/make/validate-lint [hunk 1]
```diff
@@ -76,6 +76,7 @@ packages=(
 	pkg/pubsub
 	pkg/random
 	pkg/reexec
+  pkg/registrar
```
Comment (LK4D4) [inline hack/make/validate-lint]: something with formatting

**Comment by LK4D4** [comment]
Looks cool
windows failures might be related though

**Comment by tiborvass** [comment]
Wow this is epic. Thanks so much @cpuguy83 !

Is there any undesirable side effect you can think of?

**Comment by tianon** [comment]
Maybe an obvious question, but does this mean that links won't survive a daemon restart, or were they still stored on-disk somewhere else besides SQLite?

**Comment by cpuguy83** [comment]
Adding some context this from IRC convo:
Yes there is an issue here:
Currently `hostConfig.Links` is whipped out the first time it is processed.
So the only place links info is stored currently is in the sqlite db... so we can't just remove it here, it will require some migration code to insert links back into containers' hostconfig from sqlite.

**Comment by swernli** [comment]
Confirmed that this works with Windows containers (both existing images after updating just docker.exe and on a fresh install).  Not sure why Windows Jenkins run is failing, but it works fine in manual tests.

LGTM

**Comment by icecrime** [comment]
Ping @mavenugo @mrjana: can you please look into this in light of your plans regarding links deprecation (or not)?

**Comment by thaJeztah** [comment]
ping @mavenugo @mrjana could you have a look at https://github.com/docker/docker/pull/16032#issuecomment-146645967 now that 1.9 has been release (and you hopefully have a bit more time :heart:)

**Comment by calavera** [comment]
we cannot remove links from one version to the other. 

If we're really going to remove links we should deprecated them in 1.10 to completely remove sqlite3 and the code related in 1.12.

If we're going to keep links as they are, we can make the transition to in-memory links in 1.10, loading from sqlite to memory in that version, and remove the dependency in 1.11.

If we're going to keep links with a different implementation under the hood, we should make sure we remove sqlite from the dependencies.

**Comment by cpuguy83** [comment]
@calavera unfortunately we are stuck with sqlite in any case since `HostConfig.Links` is cleared out once read, currently. So the only place the link info is stored is in sqlite... so basically as long as we want to support upgrading from \<docker version with sqlite links\> --> \<docker version with something else\> it'll need to stay around.

**Comment by thaJeztah** [comment]
@cpuguy83 should we close this for now or is this still an option? I see a discussion around removal of links, which is not yet clear when that'll happen

**Comment by cpuguy83** [comment]
@thaJeztah The overall goal is still valid, don't use sqlite for storing things on disk which we can build at runtime easily enough.
This should ultimately be much more efficient up to a certain point.

**Comment by LK4D4** [comment]
@cpuguy83 need rebase
Did you add migration logic?

**Comment by cpuguy83** [comment]
@LK4D4 Not yet.
Rebased for now, will work on migration.

**Comment by cpuguy83** [comment]
Added migration logic and test for migration.

**Comment by cpuguy83** [comment]
Errors look legit, investigating.

**Comment by cpuguy83** [comment]
Ok, all fixed.

**Comment by tiborvass** [comment]
@cpuguy83 I can't tell if the in-memory state of links is restored on daemon boot. Can you confirm?
Otherwise, design +1 with migration.

/cc @mavenugo

**Comment by cpuguy83** [comment]
It is, and there are a couple of tests to ensure it.

**Comment by tiborvass** [comment]
cool thanks, then this needs a rebase :)

**Comment by cpuguy83** [comment]
lol -- love it.

**Comment by cpuguy83** [comment]
omg, 3rd rebase today.

**Comment by cpuguy83** [comment]
Ok, this is ready.

**Comment by tiborvass** [comment]
sorry @cpuguy83 :( needs rebase again, i want this

**Comment by cpuguy83** [comment]
Rebased

**Comment by thaJeztah** [comment]
sorry @cpuguy83, looks like this needs another rebase :cry:

**Comment by cpuguy83** [comment]
Rebased

**Comment by cpuguy83** [comment]
Rebased and addressed comments.

**Comment by calavera** [comment]
LGTM

**Comment by calavera** [comment]
@cpuguy83 you might want to fix the message in your latest commit `s/hostConifg/hostConfig` :laughing:

**Comment by cpuguy83** [comment]
It's how I spell it quite frequently :)

**Comment by cpuguy83** [comment]
Fixed commit message.

**Comment by mavenugo** [comment]
@cpuguy83 @icecrime @thaJeztah as discussed in the Alias proposal, we are not deprecating links. Infact the fix for retaining the links in hostConfig is much appreciated due to the nature of the decision of reusing links as the way to provide alias support for containers in user-defined networks.

I haven't spent time testing this PR manually yet. But the code LGTM.

Given the nature of changes and potential of the impact, it is better to do some additional migration tests (in addition to all the tests that is in Integration-cli).

@cpuguy83 can you add please some test-cases and results to make sure we are not regressing particularly upgrade scenarios.

**Comment by cpuguy83** [comment]
Rebased.

@mavenugo Do you have any suggestions for migration tests beyond what's in https://github.com/cpuguy83/docker/blob/remove_sqlite_dep/daemon/links_test.go?

**Comment by thaJeztah** [comment]
darn, @cpuguy83 needs a rebase again :cry: this one is mentioned in the review-session e-mail so hopefully we can get this in asap to safe you from rebase hell

**Comment by mavenugo** [comment]
@cpuguy83 sorry. missed your last comment. I was thinking of cases where user upgrades to 1.10, downgrades to 1.9 or earlier and again upgrades to 1.10. I know your change to retaining the hostConfig.Links will come to help us. But just to make sure we have a test will give us a peace of mind.

**Comment by cpuguy83** [comment]
A graceful downgrade is not something that I think is really possible here, and is also not something we've ever really guaranteed.

**Comment by calavera** [comment]
I might be wrong but, the graceful downgrade is guaranteed unless we remove `linkgraph.db`. Obviously, if people add new containers to 1.10 and then downgrade those names will be lost. But old container names will be preserved in the database as long as the database file exists. That's guarantee enough for me.

**Comment by cpuguy83** [comment]
That's fine, I updated to make it not remove the db.

**Comment by cpuguy83** [comment]
All green

File: daemon/daemon.go [hunk 2]
```diff
@@ -332,63 +326,65 @@ func (daemon *Daemon) restore() error {
...
+		}
+	}
 
-			if c, ok := containers[e.ID()]; ok {
-				c.registered = true
-			}
+	// migrate any legacy links from sqlite
+	linkdbFile := filepath.Join(daemon.root, "linkgraph.db")
+	var legacyLinkDB *graphdb.Database
+	if _, err = os.Stat(linkdbFile); err == nil {
```
Comment (tiborvass) [inline daemon/daemon.go]: @cpuguy83 couldn't we do the `container.HostConfig == nil || container.HostConfig.Links != nil` check in the `containers` loop above to detect if legacyLinkDB should be set at all?

Comment (cpuguy83) [inline daemon/daemon.go]: Yep!

File: daemon/daemon.go [hunk 3]
```diff
@@ -332,63 +326,65 @@ func (daemon *Daemon) restore() error {
...
-			c.container.Name, err = daemon.generateNewName(c.container.ID)
-			if err != nil {
-				logrus.Debugf("Setting default id - %s", err)
-			}
-			if err := daemon.registerName(c.container); err != nil {
-				logrus.Errorf("Failed to register container %s: %s", c.container.ID, err)
-				continue
+		if legacyLinkDB != nil {
+			if err := daemon.migrateLegacySqliteLinks(legacyLinkDB, c); err != nil {
+				legacyLinkDB.Close()
```
Comment (tiborvass) [inline daemon/daemon.go]: no need for this, we already deferred it above.

**Comment by tiborvass** [comment]
LGTM

**Comment by LK4D4** [comment]
LGTM
let's test this in real world :)

**Comment by cpuguy83** [comment]
ping @ibuildthecloud aka, demolisher of code, hopes, and dreams.

**Comment by thaJeztah** [comment]
Yay, merged! Thanks for being so patient @cpuguy83 :metal:

**Comment by vincentwoo** [comment]
Is this the candidate fix to #17720? If this is on experimental I'm down to see if it'll still lockup on my production boxes.

**Comment by cpuguy83** [comment]
@vincentwoo This definitely speeds up some pieces of `docker ps`, and container creation. I'm not sure that it would fix speed reductions over time.

**Comment by vincentwoo** [comment]
I can give it a run. Is it on https://experimental.docker.com? I have a hard time figuring out what revision that thing is at.

**Comment by cpuguy83** [comment]
This is in stable.

**Comment by vincentwoo** [comment]
Right, but there's not binaries out against it yet, right? Slated for 1.10?

**Comment by cpuguy83** [comment]
https://master.dockerproject.org

**Comment by vincentwoo** [comment]
Niiiice.

**Pull request merged**